src/main/scala/org/virtuslab/internal/load/parse/ParserImpl.scala

```diff
@@ -1,34 +1,152 @@
 package org.virtuslab.internal.load.parse
 
 import scala.annotation.tailrec
-import org.virtuslab.internal.load.reader.YamlReader
+import org.virtuslab.internal.load.reader.Reader
 import org.virtuslab.internal.YamlError
 import org.virtuslab.internal.load.reader.token.Token
+import org.virtuslab.internal.load.reader.token.ScalarStyle
 
+private enum Production:
+  case ParseStreamStart
+  case ParseStreamEnd
+  case ParseDocumentStart
+  case ParseDocumentOptStart
+  case ParseDocumentEnd
+  case ParseMappingStart
+  case ParseMappingEnd
+  case ParseSequenceStart
+  case ParseSequenceEnd
+  case ParseNode
+  case ParseNodeOpt
+  case ParseKey
+  case ParseOptKey
+  case ParseScalar
+  case ParseSequenceEntryOpt
+
+/**
+ * ParserImpl is using following productions:
+ * 
+ * ParseStreamStart      ::= <ParseDocumentStart> <ParseDocumentOptStart> ParseStreamEnd
+ * ParseDocumentStart    ::= <ParseNode> ParseDocumentEnd
+ * ParseDocumentOptStart ::= epsilon | <ParseDocumentStart> <ParseDocumentOptStart>
+ * ParseNode             ::= <ParseMappingStart> | <ParseSequenceStart> | ParseScalar
+ * ParseMappingStart     ::= ParseKey <ParseNode> <ParseOptKey> ParseMappingEnd
+ * ParseSequenceStart    ::= <ParseNode> <ParseSequenceEntryOpt> ParseSequenceEnd
+ * ParseOptKey           ::= epsilon | ParseKey <ParseNode> <ParseOptKey>
+ * ParseSequenceEntryOpt ::= epsilon | <ParseNode> ParseSequenceEntryOpt
+*/
 object ParserImpl extends Parser:
-  def getEvents(
-      in: YamlReader
-  ): Either[YamlError, List[Event]] = {
+  import Production.*
+
+  def getEvents(in: Reader): Either[YamlError, List[Event]] =
     @tailrec
-    def loop(acc: List[Event]): List[Event] = {
-      val (event) = ParserImpl.getNextEvent(in)
-      if event != Event.StreamEnd then loop(acc :+ event)
+    def loop(acc: List[Event], productionStack: List[Production]): List[Event] = {
+      val (event, next) = ParserImpl.getNextEvent(in, productionStack)
+      if event != Event.StreamEnd then loop(acc :+ event, next)
       else acc :+ event
```

Comment:
Can there be a case when `event == Event.StreamEnd` but next isn't empty?

Reply:
No, we expect that `Event.StreamEnd` is the last event, and there are no more productions at stack.